### PR TITLE
Add option to compare outputs as lists of real numbers #1540

### PR DESCRIPTION
**Pull request by ronryv**
Output is considered to be correct iff for each corresponding pair of real numbers the absolute or relative difference is smaller or equal to 10^-6.

### CODE REVIEW

File: cms/grading/steps/realprecision.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+def _extract_fixed_decimals(stream: typing.BinaryIO) -> list[float]:
+    """Extract and parse all fixed-format decimal tokens from a binary stream."""
+    data = stream.read()
+    nums: list[float] = []
+    for m in _FIXED_DEC_RE.findall(data):
+        v = _parse_fixed(m)
+        if v is not None:
+            nums.append(v)
+    return nums
```
[the remark this answers is not in the document]
Comment (ronryv) [reply cms/grading/steps/realprecision.py]: I wanted to allow this feature for outputs with more than one real precision number as well, and it was the simplest implementation in order to not differentiate between whitespaces and other non fixed-format decimal characters, but we can decide not to allow it if it's problematic

File: cms/grading/tasktypes/Batch.py [hunk 2]
```diff
@@ -181,6 +227,40 @@ def _uses_grader(self) -> bool:
     def _uses_checker(self) -> bool:
         return self.output_eval == self.OUTPUT_EVAL_CHECKER
 
+    def _uses_realprecision(self) -> bool:
+        return self.output_eval == self.OUTPUT_EVAL_REALPREC
+
+    def validate_parameters(self):
```
[the remark this answers is not in the document]
Comment (ronryv) [reply cms/grading/tasktypes/Batch.py]: The validator in the parent class tests the parameters list is of the length self.ACCEPTED_PARAMETERS, so if we want to maintain backwards compatibility and accept both 3 or 4 parameters, we should override it

(parent class validator:
```
    def validate_parameters(self):
        """Validate the parameters syntactically.

        raise (ValueError): if the parameters are malformed.

        """
        if not isinstance(self.parameters, list):
            raise ValueError(
                "Task type parameters for %s are not a list" % self.__class__)

        if len(self.parameters) != len(self.ACCEPTED_PARAMETERS):

```
)

File: cms/grading/steps/realprecision.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        if num_val is not None:
+            tokens.append(('number', num_val))
+        
+        pos = match.end()
+    
+    if pos < len(data):
+        text = data[pos:]
+        canonical = _white_diff_canonicalize(text)
+        if len(canonical) > 0:
+            tokens.append(('text', canonical))
```
[the remark this answers is not in the document]
Comment (ronryv) [reply cms/grading/steps/realprecision.py]: Devin:
Thanks for the suggestion! However, the proposed one-liner is **not equivalent** to the current implementation. Here are the key differences:

## 1. Empty text tokens
The current code intentionally drops text chunks that canonicalize to empty bytes (whitespace-only segments). The suggested version would emit `('text', b'')` tokens because it filters `x != b''` *before* canonicalization, not after.

**Example:** Input `b"1  2"` (two numbers separated by whitespace)
- Current: `[('number', 1.0), ('number', 2.0)]` ✓
- Suggested: `[('number', 1.0), ('text', b''), ('number', 2.0)]` ❌

This would break the comparison logic downstream, which expects text tokens to have content.

## 2. Parse failures
The current code only appends a number token if `_parse_fixed()` returns a float. The suggested version always emits `('number', _parse_fixed(x))`, which could be `('number', None)` on parse failure, breaking downstream code that expects floats.

## 3. Regex duplication
The suggestion duplicates the regex pattern string instead of reusing `_FIXED_DEC_RE`. This creates a maintenance risk if the pattern needs to change.

## 4. Bytes vs string literal
The original suggestion used `x != ''` (string) instead of `x != b''` (bytes), which wouldn't filter anything since `data` is bytes.

## Test results
I tested both implementations with various edge cases:

```python
Input: b'1  2'           # whitespace between numbers
  Current:   [('number', 1.0), ('number', 2.0)]
  Suggested: [('number', 1.0), ('text', b''), ('number', 2.0)]  ❌

Input: b'  1'            # leading whitespace
  Current:   [('number', 1.0)]
  Suggested: [('text', b''), ('number', 1.0)]  ❌

Input: b'a\t\n b 3 .5'   # mixed whitespace
  Current:   [('text', b'a b'), ('number', 3.0), ('number', 0.5)]
  Suggested: [('text', b'a b'), ('number', 3.0), ('text', b''), ('number', 0.5)]  ❌
```

## Recommendation
I'd keep the current `finditer`-based implementation. It's clearer, avoids regex duplication, and correctly handles the edge cases. The explicit loop makes the filtering logic (dropping empty canonicalized text, skipping failed parses) more obvious and maintainable.
